### PR TITLE
Signed enclave responses

### PR DESCRIPTION
**Pull request by podcastinator**
This PR adds signatures on enclave output, binding the output to the submitted command using the nonce / sequence number.

### CODE REVIEW

File: python-package/securexgboost/remote_server.py [hunk 1]
```diff
@@ -210,20 +212,23 @@ def rpc_XGBoosterSetParam(self, request, context):
         Set booster parameter
         """
         try:
-            _ = self._synchronize(remote_api.XGBoosterSetParam, request)
-            return remote_pb2.Status(status=0)
+            sig, sig_len = self._synchronize(remote_api.XGBoosterSetParam, request)
+            sig_proto = pointer_to_proto(sig, sig_len)
```
Comment (chester-leung) [inline python-package/securexgboost/remote_server.py:216]: This line seems to be repeated in various places. Can we call `pointer_to_proto(sig, sig_len)` before returning out of `self._synchronize`? Maybe we can call it in `invoke()` after running `self._func` before setting `self._ret`?

File: python-package/securexgboost/core.py [hunk 2]
```diff
@@ -3056,10 +3201,38 @@ def sign_data(key, data, data_size):
...
+
+    pem_key = globals()["enclave_pk"]
+    pem_key_len = globals()["enclave_pk_size"]
+    # Verify signature
+    _check_call(_LIB.verify_signature(pem_key, pem_key_len, arr, size, sig, sig_len))
+
+    globals()["nonce_ctr"] += 1
+
+
+def create_enclave_signature(args):
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Is my understanding correct that this function is called to create a client-side signature over the arguments and nonce? If so, can we rename the function? The name is a bit confusing as it implies creating a signature in the enclave. Maybe we can rename to `create_client_signature`?

File: python-package/securexgboost/core.py [hunk 3]
```diff
@@ -3045,8 +3187,11 @@ def sign_data(key, data, data_size):
     # Cast data : proto.NDArray to pointer to pass into C++ sign_data() function
     if isinstance(data, str):
         data = c_str(data)
-    # FIXME use positive check instead of negative check
-    elif not isinstance(data, ctypes.c_char_p):
+    elif isinstance(data, ctypes.Array) and (data._type_ is ctypes.c_char):
+        pass
+    else:
+        # FIXME error handling for other types
+        print(type(data))
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Remove this print statement?

File: python-package/securexgboost/core.py [hunk 4]
```diff
@@ -17,8 +17,9 @@
 import warnings
 
 import grpc
-from .rpc import remote_pb2
+from .rpc import remote_pb2, ndarray_pb2
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Do we need to import `ndarray_pb2`?

File: python-package/securexgboost/remote_server.py (hunk 1 above)
Comment (podcastinator) [reply python-package/securexgboost/remote_server.py:216]: The trouble is that `self._synchronize` simply returns whatever is returned by the invoked function, and the number of values returned by each function are different. So, `invoke` has no knowledge of which returned value is the signature.

**Review by chester-leung — APPROVED** [review]

**Pull request merged**